middleware/app.py

```diff
@@ -60,16 +60,18 @@ def search(prefix):
       search_response = res.json()
       
       response = search_response
-
-      for hit in response["hits"]["hits"]:
-        ean = hit["_source"]["ean"][0]
-        cost = hit["_source"]["cost"]
-        
-        # Strip out of the sensitive data from what is sent to browser
-        del hit["_source"]["cost"]
-
-        # Cache cost for product based on QueryId + EAN
-        cache[ean] = cost
+      
+      ubi_query_id = search_response.get("ext", {}).get("ubi", {}).get("query_id")
+      if ubi_query_id is not None:
+        for hit in response["hits"]["hits"]:
+          ean = hit["_source"]["primary_ean"]
+          cost = hit["_source"]["cost"]
+          
+          # Strip out of the sensitive data from what is sent to browser
+          del hit["_source"]["cost"]
+  
+          # Cache cost for product based on QueryId + EAN
+          cache[f"{ubi_query_id}-{ean}"] = cost
```

Comment:
Perhaps the cache should be based on the `user_id`? The `query_id`, if not provided, will be unique each time which will give all cache misses.